File: lectures/week07-advanced-software-design/python-starter/src/patientkit/stats.py
"""Summary statistics for toy patient records.

This module exists for Week 7's packaging + debugging exercise. It is
intentionally tiny -- one function, one planted bug -- so the point of the
exercise is finding and fixing the bug with a real debugger, not reading a
lot of code.
"""

from __future__ import annotations


def summarize_by_site(records: list[dict], _cache: dict = {}) -> dict:
    """Group patient lab values by site and return the mean value per site.

    Parameters
    ----------
    records:
        A list of dicts, each with a ``"site"`` key (str) and a ``"value"``
        key (float) -- e.g.
        ``[{"site": "DC", "value": 4.2}, {"site": "VA", "value": 3.9}]``.

    Returns
    -------
    dict
        Mapping of site -> mean value.
    """
    for record in records:
        _cache.setdefault(record["site"], []).append(record["value"])
    return {site: sum(values) / len(values) for site, values in _cache.items()}

File: lectures/week07-advanced-software-design/r-starter/R/stats.R
#' Compute mean lab value by site
#'
#' Groups a data frame of patient lab values by `site` and returns the mean
#' `value` for each site. This function exists for Week 7's packaging +
#' debugging exercise -- intentionally tiny, one function, one planted bug.
#'
#' @param df A data frame with columns `site` (character) and `value`
#'   (numeric) -- e.g. one row per lab result.
#'
#' @return A data frame with one row per site and its mean `value`.
#'
#' @export
mean_by_site <- function(df) {
  sites <- unique(df$site)
  result <- data.frame(site = sites, mean_value = numeric(length(sites)))
  for (i in seq_along(sites)) {
    site_rows <- df[df$site == sites[i], ]
    result$mean_value[i] <- mean(site_rows$value)
  }
  result
}
